src/huggingface_hub/repocard.py

```diff
@@ -99,3 +104,153 @@ def metadata_eval_result(
+            If set to `True` an existing field can be overwritten, otherwise
+            attempting to overwrite an existing field will cause an error.
+        token (`str`, *optional*):
+            The Hugging Face authentication token
+    """
+
+    filepath = hf_hub_download(
+        repo_id,
+        filename="README.md",
+        repo_type=repo_type,
+        use_auth_token=token,
+        force_download=True,
+    )
+    existing_metadata = metadata_load(filepath)
+
+    for key in metadata:
+        # update all fields except model index
+        if key != "model-index":
+            if key in existing_metadata and not overwrite:
+                if existing_metadata[key] != metadata[key]:
+                    raise ValueError(
+                        f"""You passed a new value for the existing meta data field '{key}'. Set `overwrite=True` to overwrite existing metadata."""
+                    )
+            else:
+                existing_metadata[key] = metadata[key]
+        # update model index containing the evaluation results
+        else:
+            if "model-index" not in existing_metadata:
+                existing_metadata["model-index"] = metadata["model-index"]
+            else:
+                # the model-index contains a list of results as used by PwC but only has one element thus we take the first one
+                existing_metadata["model-index"][0][
+                    "results"
+                ] = _update_metadata_model_index(
+                    existing_metadata["model-index"][0]["results"],
+                    metadata["model-index"][0]["results"],
+                    overwrite=overwrite,
+                )
+
+    # save and push to hub
+    metadata_save(filepath, existing_metadata)
+
+    HfApi().upload_file(
+        path_or_fileobj=filepath,
+        path_in_repo="README.md",
+        repo_id=repo_id,
+        repo_type=repo_type,
+        identical_ok=True,
+        token=token,
+    )
+
+
+def _update_metadata_model_index(existing_results, new_results, overwrite=False):
+    """
+    Updates the model-index fields in the metadata. If results with same unique
+    features exist they are updated, else a new result is appended. Updating existing
+    values is only possible if `overwrite=True`.
+
+    Args:
+        new_metrics (`list`):
+            List of new metadata results
+        existing_metrics (`list`):
+            List of existing metadata results
+        overwrite (`bool`, *optional*, defaults to `False`):
+            If set to `True` an existing metric values can be overwritten, otherwise
+            attempting to overwrite an existing field will cause an error.
+
+    Returns:
+        `list`: List of updated metadata results
+    """
+    for new_result in new_results:
+        try:
+            existing_result = next(
+                x
+                for x in existing_results
+                if all(x[feat] == new_result[feat] for feat in UNIQUE_RESULT_FEATURES)
```

Comment:
Since we pull the metadata from the Hub can't we assume that it is valid? In your experiment pushing invalid metadata was rejected, right? So it should not be there in the first place if it is invalid.

Reply:
Hmm @julien-c WDYT? https://huggingface.co/osanseviero/llama-horse-zebra 

The result has task and metrics, but no associated dataset. This is not rejected by the server, and the evaluation results are nicely shown, with the only con that there is no associated dataset, so there's no leaderboard.

IMO this is still valid metadata which is just incomplete. Most spaCy models are like this https://huggingface.co/models?library=spacy

Reply:
yes we try to still display the eval results in this case even though it's not perfect